Update community info page

darrenvong opened the pull request:
<!---
Please describe the pull request below.
If it closes an issue, make sure to write "closes #1234"
If there is an issue but it isn't completely closed, still refer to the issue number, eg. "part of #1234"
--->
Closes #927

**Desktop (community info page)**
With new "edit" link near the top right underneath the tabs if user has permission. Same for mobile.
![Screenshot 2021-06-12 at 02 37 36](https://user-images.githubusercontent.com/13669362/121761407-73afdb80-cb27-11eb-936d-4d40197ff3a2.png)

**Mobile (edit page)**
![Screenshot 2021-06-12 at 02 37 21](https://user-images.githubusercontent.com/13669362/121761413-7d394380-cb27-11eb-8d5d-e62b980208a6.png)


**Desktop (edit page)**
![Screenshot 2021-06-12 at 02 37 06](https://user-images.githubusercontent.com/13669362/121761368-3fd4b600-cb27-11eb-897b-ca574be9ba6c.png)

**Mobile (edit page)**
![Screenshot 2021-06-12 at 02 36 53](https://user-images.githubusercontent.com/13669362/121761375-46632d80-cb27-11eb-84b0-fd2a96abbc90.png)

<!---
Checklists - you can remove one that is not applicable (ie. remove backend checklist if you only worked on frontend)
If you need help with any of these, please ask :)
--->

**Frontend checklist**
- [x] Formatted my code with `yarn format && yarn lint --fix`
- [x] There are no warnings from `yarn lint`
- [x] There are no console warnings when running the app
- [x] Added any new components to storybook
- [x] Added tests where relevant
- [x] All tests pass
- [x] Clicked around my changes running locally and it works
- [ ] Checked Desktop, Mobile and Tablet screen sizes


<!---
Remember to request review from couchers-org/frontend, couchers-org/@backend or an individual.
Once your code is approved, remember to merge it if you have write access
--->

darrenvong committed:
feat: add edit community page

darrenvong commented on app/frontend/src/features/communities/EditCommunityPage.tsx:
```diff
@@ -0,0 +1,145 @@
+      service.pages.updatePage({ content, pageId: +pageId }),
+    {
+      onSuccess(newPageData, { communityId, slug }) {
+        queryClient.setQueryData<Community.AsObject | undefined>(
```
Optimistic update in `onSuccess` for some reason doesn't work here and I'm not quite sure what I've done wrong compared to other similar queries... on slow network, even the `history.push` redirect you can see the content is rolling back to the old state in the editor before it redirects you back to the actual page with the new content displaying correctly.

lucaslcode replied:
It works if this is added to `MarkdownInput.tsx`

```
const {
    ref: fieldRef,
    onBlur: fieldOnBlur,
    onChange: fieldOnChange,
    value,
  } = field;

  useEffect(() => (initialDefaultValue.current = value), [value]);
```

The useEffect below this that creates the ToastEditor is triggered when you press the update button, which destroys the editor and re-creates it with the original default value.
It would be more correct to stop `useEffect` being triggered, but I have no idea why it is... I tried memoizing stuff to no avail

darrenvong replied:
Good catch! It looks like `onBlur` and `onChange` from `useController` doesn't have a stable identity and gets re-created on every re-render, and we can't really memoise with `useCallback` since the exhaustive dep rule will just ask `fieldonBlur`/`fieldonChange` to be a dependency which will lead to the same problem as in that `useEffect`...

The other workaround I came up with is storing `fieldOnBlur`/`fieldOnChange` in refs and call that in the `useEffect` that creates the ToastEditor instead. Since neither `fieldOnBlur` nor `fieldOnChange` depends on state, it's probably ok?

darrenvong committed:
feat: store onBlur/onChange from useController in refs (MarkdownInput)

The identity in those functions are annoyingly unstable. Storing them in
refs will stabilise their identity, so that it prevents the effect from
wrongly run more than once and destroying the editor instance.

This should be ok since neither functions read from state or props from
the MarkdownInput component

darrenvong committed:
feat: add tests and storybook entries for edit community page

darrenvong commented:
@lucaslcode This basic version of edit page should be ready to be reviewed now

lucaslcode commented on app/frontend/src/components/MarkdownInput.tsx:
```diff
@@ -61,16 +61,22 @@ export default function MarkdownInput({
+  const { ref: fieldRef } = field;
 
   const rootEl = useRef<HTMLDivElement>(null);
+
```
Comment here why this was necessary

lucaslcode commented on app/frontend/src/features/communities/constants.ts:
```diff
@@ -16,12 +16,15 @@ export const COMMENT = "Comment";
 export const DISCUSSIONS_EMPTY_STATE = "No discussions at the moment.";
 export const DISCUSSIONS_LABEL = "Discussions";
 export const DISCUSSIONS_TITLE = "Discussions";
+export const EDIT_COMMUNITY_PAGE = "Edit community page";
```
I wonder if this should be Edit local info? Can probably get away with changing this only and not all the components 😆

lucaslcode commented on app/frontend/src/routes.ts:
```diff
@@ -127,3 +127,9 @@ export const routeToCommunity = (
   slug: string,
   page?: CommunityTab
 ) => `${communityBaseRoute}/${id}/${slug}${page ? `/${page}` : ""}`;
+export const editCommunityPageRoute = `${communityRoute}/edit`;
```
With this setup, when editing the page, the "Overview" tab is selected. Would be better if local info was.

darrenvong replied:
That's not the case if you look at the screenshot, because `routeToEditCommunityPage` sets the `page` param in the route to "info", so if you click from the link, it works fine.

~~However it doesn't stop people from going `/community/1/world/some-gibberish/edit` and still hitting the edit info page form. Hmm maybe rendering the "not found page" in those cases instead is best~~ Lol ignore this, literally making the route more specific to match on `/community/<id>/<slug>/info/edit` solved this

lucaslcode reviewed (changes requested):
Can you also do one more thing - remove the description from under the title and use that as the General Info section instead (and render it as markdown instead of stripping). 

I don't think using the local info page excerpt will look good, and community mods can just make sure the markdown for the description is appropriate.

darrenvong committed:
fix: match edit community info page route more specifically

darrenvong committed:
refactor: rename edit community page -> edit community info page

lucaslcode commented on app/frontend/src/features/communities/CommunityPage/CommunityPage.tsx:
```diff
@@ -11,9 +11,7 @@ import InfoPageSection from "./InfoPageSection";
-    ...theme.typography.h1Large,
+    marginTop: theme.spacing(3),
   },
   description: {
```
Remove

lucaslcode commented on app/frontend/src/features/communities/CommunityPage/InfoPageSection.tsx:
```diff
@@ -1,12 +1,12 @@
 import makeStyles from "utils/makeStyles";
 
 import { GENERAL_INFORMATION, SEE_MORE_INFORMATION } from "../constants";
-import getContentSummary from "../getContentSummary";
```
Is this still used anywhere?

darrenvong replied:
Yeah looks like this might be unnecessary now as `PlaceCard` uses the `LineEllipsis` for doing the ...

I'll remove it!

darrenvong replied:
Oh wait, discussion still uses it

lucaslcode reviewed (changes requested):
Nits

darrenvong committed:
fix: remove unused community description styles
